src/include/gurt/debug.h

```diff
@@ -197,6 +195,32 @@ extern void (*d_alt_assert)(const int, const char*, const char*, const int);
 #define D_TRACE_EMIT(ptr, fmt, ...)	\
 	D_TRACE_DEBUG(DLOG_EMIT, ptr, fmt, ## __VA_ARGS__)
 
+/* Alternative versions of D_ERROR and D_WARN.
+ *
+ * These take a rc value and possibly a descriptor so we are able to better control the output.
+ */
+#define DHL_ERROR(_desc, _rc, _fmt, ...)                                                           \
+	do {                                                                                       \
+		_D_DEBUG(_D_TRACE_NOCHECK, DLOG_ERR, (_desc), _fmt ": " DF_RC "\n", ##__VA_ARGS__, \
+			 DP_RC(_rc));                                                              \
+	} while (0)
+
+#define DL_ERROR(_rc, _fmt, ...)                                                                   \
+	do {                                                                                       \
+		D_DEBUG(DLOG_ERR, _fmt ": " DF_RC "\n", ##__VA_ARGS__, DP_RC(_rc));                \
+	} while (0)
+
+#define DHL_WARN(_desc, _rc, _fmt, ...)                                                            \
+	do {                                                                                       \
+		_D_DEBUG(_D_TRACE_NOCHECK, DLOG_WARN, (_desc), _fmt ": " DF_RC "\n",               \
```

Comment:
What is the intention of DHL * ?

Reply:
I can't remember how the names came about but it adds a descriptor so you can trace error/debug messages to where they originate from.  The idea is once you get a DL_ERROR call for a descriptor then you can trace back and highlight all the debug messages from it.

Two problems with this though, it's not possible to automatically find the right pointer to use as `desc` so code will need updating manually and the above only works if debug messages are updated as well as error messages.